// mcpster/dist/server.js
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { connectStdio } from './transport/stdio.js';
import { connectHttp } from './transport/http.js';
import { registerTool } from './tool.js';
import { registerResource } from './resource.js';
import { registerPrompt } from './prompt.js';
import { applySetup } from './setup.js';
class McpsterServerImpl {
    sdk;
    config;
    toolNames = [];
    _stop;
    constructor(config) {
        this.config = config;
        this.sdk = new McpServer({
            name: config.name,
            version: config.version,
        });
    }
    defineTool(def) {
        registerTool(this.sdk, def);
        this.toolNames.push(def.name);
        return this;
    }
    defineResource(def) {
        registerResource(this.sdk, def);
        return this;
    }
    definePrompt(def) {
        registerPrompt(this.sdk, def);
        return this;
    }
    async setup(options) {
        const isHttp = this.config.transport === 'http';
        applySetup(this.config.name, this.toolNames, {
            register: !isHttp,
            ...options,
        });
        return this;
    }
    async start() {
        if (this.config.transport === 'http') {
            this._stop = await connectHttp(this.sdk, this.config.http);
        }
        else {
            this._stop = await connectStdio(this.sdk);
        }
    }
    async stop() {
        await this._stop?.();
    }
}
export function createServer(config) {
    return new McpsterServerImpl(config);
}
