// mcpster/dist/setup.js
import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'fs';
import { homedir } from 'os';
import { join, resolve } from 'path';
function claudeJsonPath() {
    return resolve(homedir(), '.claude.json');
}
function readClaudeJson() {
    try {
        return JSON.parse(readFileSync(claudeJsonPath(), 'utf8'));
    }
    catch {
        return {};
    }
}
function writeClaudeJson(data) {
    writeFileSync(claudeJsonPath(), JSON.stringify(data, null, 2) + '\n', 'utf8');
}
function applyRegister(serverName, projectPath, startArgs = []) {
    const claudeDir = resolve(projectPath, '.claude');
    const settingsPath = resolve(claudeDir, 'settings.json');
    let settings = {};
    if (existsSync(settingsPath)) {
        try {
            settings = JSON.parse(readFileSync(settingsPath, 'utf8'));
        }
        catch {
            // start fresh if file is malformed
        }
    }
    const mcpServers = (settings.mcpServers ?? {});
    mcpServers[serverName] = {
        command: 'node',
        args: [process.argv[1], ...startArgs],
    };
    settings.mcpServers = mcpServers;
    try {
        mkdirSync(claudeDir, { recursive: true });
        writeFileSync(settingsPath, JSON.stringify(settings, null, 2) + '\n', 'utf8');
        console.error(`[mcpster] registered "${serverName}" in .claude/settings.json`);
    }
    catch {
        // silently skip — cwd may not be a writable project directory (e.g. launched from /)
    }
    // Also register at user scope in ~/.claude.json so the server is available from any directory
    const claudeJson = readClaudeJson();
    const userMcpServers = (claudeJson.mcpServers ?? {});
    userMcpServers[serverName] = {
        type: 'stdio',
        command: 'node',
        args: [process.argv[1], ...startArgs],
        env: {},
    };
    claudeJson.mcpServers = userMcpServers;
    writeClaudeJson(claudeJson);
    console.error(`[mcpster] registered "${serverName}" at user scope in ~/.claude.json`);
}
function desktopConfigPaths() {
    const home = homedir();
    if (process.platform === 'win32') {
        const appData = process.env.APPDATA ?? join(home, 'AppData', 'Roaming');
        return [join(appData, 'Claude', 'claude_desktop_config.json')];
    }
    return [join(home, 'Library', 'Application Support', 'Claude', 'claude_desktop_config.json')];
}
function patchJsonFile(filePath, patcher, label) {
    try {
        if (!existsSync(filePath))
            return;
        const config = JSON.parse(readFileSync(filePath, 'utf8'));
        patcher(config);
        writeFileSync(filePath, JSON.stringify(config, null, 2) + '\n', 'utf8');
        console.log(`[mcpster] registered "${label}"`);
    }
    catch {
        // silently skip — user may have a different stack
    }
}
export function installServer(serverName, serverPath, env) {
    const home = homedir();
    patchJsonFile(join(home, '.claude.json'), config => {
        const servers = (config.mcpServers ?? {});
        servers[serverName] = { type: 'stdio', command: 'node', args: [serverPath], env: env ?? {} };
        config.mcpServers = servers;
    }, `${serverName} in Claude Code (~/.claude.json)`);
    for (const configPath of desktopConfigPaths()) {
        patchJsonFile(configPath, config => {
            const servers = (config.mcpServers ?? {});
            const entry = { command: 'node', args: [serverPath] };
            if (env && Object.keys(env).length > 0)
                entry.env = env;
            servers[serverName] = entry;
            config.mcpServers = servers;
        }, `${serverName} in Claude desktop`);
    }
}
export function applySetup(serverName, toolNames, options = {}) {
    const projectPath = resolve(options.projectPath ?? process.cwd());
    const permissions = options.permissions ?? 'restrictive';
    const claude = readClaudeJson();
    const projects = (claude.projects ?? {});
    const project = projects[projectPath] ?? {};
    const allowedTools = permissions === 'permissive'
        ? toolNames.map(name => `${serverName}:${name}`)
        : [];
    projects[projectPath] = { ...project, allowedTools };
    claude.projects = projects;
    writeClaudeJson(claude);
    if (options.register !== false) {
        applyRegister(serverName, projectPath, options.startArgs ?? []);
    }
    console.error(`[mcpster] setup complete — permissions: ${permissions}`);
    if (allowedTools.length > 0) {
        console.error(`[mcpster] allowed tools: ${allowedTools.join(', ')}`);
    }
}
