// mcpster/dist/transport/http.js
import express from 'express';
import { randomUUID } from 'node:crypto';
import { readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { dirname } from 'node:path';
import { StreamableHTTPServerTransport } from '@modelcontextprotocol/sdk/server/streamableHttp.js';
import { mcpAuthRouter } from '@modelcontextprotocol/sdk/server/auth/router.js';
import { requireBearerAuth } from '@modelcontextprotocol/sdk/server/auth/middleware/bearerAuth.js';
function createClientsStore(clientsFile) {
    const clients = new Map();
    if (clientsFile) {
        try {
            const raw = readFileSync(clientsFile, 'utf8');
            const stored = JSON.parse(raw);
            for (const c of stored)
                clients.set(c.client_id, c);
        }
        catch {
            // File doesn't exist yet — start empty
        }
    }
    function persist() {
        if (!clientsFile)
            return;
        try {
            mkdirSync(dirname(clientsFile), { recursive: true });
            writeFileSync(clientsFile, JSON.stringify([...clients.values()], null, 2), 'utf8');
        }
        catch (err) {
            console.error('[mcpster] Failed to persist OAuth clients:', err);
        }
    }
    return {
        getClient: (id) => clients.get(id) ?? undefined,
        registerClient: (client) => {
            // Always treat dynamically registered clients as public (no client_secret).
            // This allows clients like Claude.ai that use PKCE and never send a secret
            // to complete the token exchange, regardless of what they sent at registration.
            const { client_secret: _secret, client_secret_expires_at: _exp, ...rest } = client;
            const full = {
                redirect_uris: [],
                ...rest,
                token_endpoint_auth_method: 'none',
                client_id: randomUUID(),
                client_id_issued_at: Math.floor(Date.now() / 1000),
            };
            clients.set(full.client_id, full);
            persist();
            return full;
        },
    };
}
// ---------------------------------------------------------------------------
// OAuth 2.1 provider (PKCE, dynamic client registration)
// ---------------------------------------------------------------------------
function createOAuthProvider(clientsFile) {
    const clientsStore = createClientsStore(clientsFile);
    const authCodes = new Map();
    const tokens = new Map();
    const refreshTokens = new Map();
    const esc = (s) => String(s ?? '').replace(/&/g, '&amp;').replace(/"/g, '&quot;').replace(/</g, '&lt;');
    return {
        get clientsStore() { return clientsStore; },
        async authorize(client, params, res) {
            const code = randomUUID();
            authCodes.set(code, {
                clientId: client.client_id,
                codeChallenge: params.codeChallenge,
                redirectUri: params.redirectUri,
                scopes: params.scopes ?? [],
            });
            res.send(`<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>Authorize — ${esc(client.client_name ?? client.client_id)}</title>
  <style>
    body { font-family: system-ui, sans-serif; max-width: 480px; margin: 80px auto; padding: 0 20px; }
    h2  { margin-bottom: 8px; }
    p   { color: #555; margin-bottom: 24px; }
    button { padding: 10px 20px; border-radius: 6px; font-size: 15px; cursor: pointer; border: 1px solid #ccc; }
    .approve { background: #0070f3; color: #fff; border-color: #0070f3; margin-right: 10px; }
  </style>
</head>
<body>
  <h2>Authorization Request</h2>
  <p><strong>${esc(client.client_name ?? client.client_id)}</strong> is requesting access to this MCP server.</p>
  <form method="POST" action="/approve">
    <input type="hidden" name="code"         value="${esc(code)}">
    <input type="hidden" name="redirect_uri" value="${esc(params.redirectUri)}">
    <input type="hidden" name="state"        value="${esc(params.state ?? '')}">
    <button class="approve" type="submit">Approve</button>
    <button type="button" onclick="history.back()">Deny</button>
  </form>
</body>
</html>`);
        },
        async challengeForAuthorizationCode(client, code) {
            const entry = authCodes.get(code);
            if (!entry || entry.clientId !== client.client_id)
                throw new Error('Invalid authorization code');
            return entry.codeChallenge;
        },
        async exchangeAuthorizationCode(client, code) {
            const entry = authCodes.get(code);
            if (!entry || entry.clientId !== client.client_id)
                throw new Error('Invalid authorization code');
            authCodes.delete(code);
            const accessToken = randomUUID();
            const refreshToken = randomUUID();
            const expiresAt = Math.floor(Date.now() / 1000) + 3600;
            tokens.set(accessToken, { token: accessToken, clientId: client.client_id, scopes: entry.scopes, expiresAt, refreshToken });
            refreshTokens.set(refreshToken, accessToken);
            return { access_token: accessToken, token_type: 'bearer', expires_in: 3600, refresh_token: refreshToken };
        },
        async exchangeRefreshToken(client, oldRefreshToken) {
            const oldAccess = refreshTokens.get(oldRefreshToken);
            if (!oldAccess)
                throw new Error('Invalid refresh token');
            const old = tokens.get(oldAccess);
            if (!old)
                throw new Error('Invalid refresh token');
            const accessToken = randomUUID();
            const newRefresh = randomUUID();
            const expiresAt = Math.floor(Date.now() / 1000) + 3600;
            tokens.delete(oldAccess);
            refreshTokens.delete(oldRefreshToken);
            tokens.set(accessToken, { token: accessToken, clientId: client.client_id, scopes: old.scopes, expiresAt, refreshToken: newRefresh });
            refreshTokens.set(newRefresh, accessToken);
            return { access_token: accessToken, token_type: 'bearer', expires_in: 3600, refresh_token: newRefresh };
        },
        async verifyAccessToken(token) {
            const info = tokens.get(token);
            if (!info)
                throw new Error('Invalid token');
            if (info.expiresAt < Math.floor(Date.now() / 1000))
                throw new Error('Token expired');
            return info;
        },
    };
}
// ---------------------------------------------------------------------------
// HTTP transport with OAuth 2.1
// ---------------------------------------------------------------------------
export async function connectHttp(server, config) {
    const port = config?.port ?? 3000;
    const path = config?.path ?? '/mcp';
    // baseUrl drives OAuth metadata (issuer, token endpoint, etc.).
    // Priority: config.baseUrl → BASE_URL env → http://localhost:<port>
    const envBaseUrl = process.env.BASE_URL;
    const rawBaseUrl = config?.baseUrl || (envBaseUrl?.startsWith('http') ? envBaseUrl : undefined) || `http://localhost:${port}`;
    const baseUrl = new URL(rawBaseUrl);
    const authEnabled = config?.auth ?? false;
    const app = express();
    // Behind a reverse proxy (Fly.io, Railway, etc.) there is exactly one hop.
    app.set('trust proxy', 1);
    app.use(express.json());
    app.use(express.urlencoded({ extended: false }));
    // Request logging
    app.use((req, _res, next) => {
        console.log(`[mcpster] ${req.method} ${req.path}`, req.body && Object.keys(req.body).length ? JSON.stringify(req.body) : '');
        next();
    });
    if (authEnabled) {
        if (!config?.clientsFile && !process.env.MCPSTER_CLIENTS_FILE) {
            console.warn('[mcpster] Warning: running HTTP transport without clientsFile — registered OAuth clients ' +
                'are stored in memory only and will be lost on restart. Set http.clientsFile in ' +
                'mcpster.config.json (e.g. "/data/oauth-clients.json") and mount a persistent volume.');
        }
        const provider = createOAuthProvider(config?.clientsFile ?? process.env.MCPSTER_CLIENTS_FILE);
        // OAuth discovery + registration + token endpoints
        app.use(mcpAuthRouter({ provider, issuerUrl: baseUrl }));
        // Approve endpoint — browser page shown during the OAuth flow
        app.post('/approve', (req, res) => {
            const { code, redirect_uri, state } = req.body;
            if (!code || !redirect_uri) {
                res.status(400).send('Missing code or redirect_uri');
                return;
            }
            const url = new URL(redirect_uri);
            url.searchParams.set('code', code);
            if (state)
                url.searchParams.set('state', state);
            res.redirect(url.toString());
        });
        // MCP endpoint — protected by bearer token
        // resourceMetadataUrl tells clients where to find OAuth discovery metadata (RFC 9728)
        const resourceMetadataUrl = new URL('/.well-known/oauth-protected-resource', baseUrl).href;
        app.all(path, requireBearerAuth({ verifier: provider, resourceMetadataUrl }), async (req, res) => {
            const transport = new StreamableHTTPServerTransport({ sessionIdGenerator: undefined, enableJsonResponse: config?.enableJsonResponse });
            try {
                await server.connect(transport);
                await transport.handleRequest(req, res, req.body);
            }
            catch {
                if (!res.headersSent)
                    res.status(500).end();
            }
            finally {
                await transport.close();
            }
        });
    }
    else {
        // No auth — MCP endpoint is open
        app.all(path, async (req, res) => {
            const transport = new StreamableHTTPServerTransport({ sessionIdGenerator: undefined, enableJsonResponse: config?.enableJsonResponse });
            try {
                await server.connect(transport);
                await transport.handleRequest(req, res, req.body);
            }
            catch {
                if (!res.headersSent)
                    res.status(500).end();
            }
            finally {
                await transport.close();
            }
        });
    }
    const httpServer = app.listen(port);
    await new Promise((resolve, reject) => {
        httpServer.once('listening', resolve);
        httpServer.once('error', reject);
    });
    return () => new Promise((resolve, reject) => {
        httpServer.close((err) => (err ? reject(err) : resolve()));
    });
}
